### PR TITLE
[WIP] Velocity-based trajectory generator for manual position control mode

### PR DESCRIPTION
**Pull request by bresch**
## Description
This PR introduces a new FlightTask that generates smooth x-y-z trajectories based on desired velocities given by the RC sticks.

Given a desired velocity setpoint v_d, the trajectory generator computes
    a time-optimal trajectory satisfying the following variable constraints:
   - maximum jerk
   - maximum acceleration
   - initial acceleration
   - initial velocity
   - final velocity

The hard constraint used to generate the equations is:
   - final acceleration = 0

The trajectory generated is made of three parts:
1. Increasing acceleration during T1 seconds
2. Constant acceleration during T2 seconds
3. Decreasing acceleration during T3 seconds

vel_traj_graph

At each loop, the optimization is computed and the trajectory is integrated to generate position, velocity, acceleration and jerk setpoints. The velocity is then used as the input of the velocity PID controller.

**Position hold continuous switch**:
The common issue when always running the position loop is that a steady-state position error will rapidly lead to a large delay from the user perspective. To avoid this, the position controller is not used during motion but is closed when the trajectory reaches low velocity and acceleration.
To open the position loop without inducing a discontinuity at the input of the velocity controller, the current velocity of the trajectory is forced to the previous input of the velocity controller at that switching point.

The algorithm uses the equations solved using [this script](https://github.com/Auterion/trajectory_generator/blob/master/velocity_trajectory_generator_symbolic.py). The trajectory generator has been tested [here](https://github.com/Auterion/trajectory_generator/blob/master/velocity_trajectory_generator.py)

**Example of flight**:
https://logs.px4.io/plot_app?log=2be0a94d-9966-44ac-ba51-7dc4ceb35c0d

### How to use it:
- Set `MPC_POS_MODE` to 3
- Tuning parameters:
`MPC_ACC_HOR_MAX` is used as the maximum acceleration/deceleration
`MPC_JERK_MIN` is used as the maximum jerk during normal operation
`MPC_JERK_MAX` is used as the maximum jerk while breaking (sticks centered)

## TODO:
- [ ] Choose consistent variable naming : 
   - acceleration -> accel
   - velocity -> vel
   - position -> pos
   - setpoint -> sp (as a suffix after pos/vel/accel)
   - use xy as a suffix after sp for a 2D vector in the XY plane
   - Example: `Vector2f accel_sp_xy`
- [ ] Add feedforward to the velocity controller to be able to use the acceleration part of the trajectory. A simple gain can then convert acceleration to thrust.
- [ ] Prevent delay induced by possible bad velocity tracking: As for the position loop, a velocity tracking error induces a delay from the user perspective (this problem has been solved so far by doing a reset [see the documentation](https://docs.px4.io/en/config_mc/advanced_mc_position_tuning.html)). A possible nicer solution would to use the same position hold trick described above and force the acceleration of the trajectory to the output of the velocity controller.

FYI @bkueng @MaEtUgR @Stifael @RomanBapst

### CODE REVIEW

File: ROMFS/px4fmu_common/init.d-posix/rcS [hunk 1]
```diff
@@ -131,6 +131,14 @@ then
 	param set MPC_HOLD_MAX_Z 2.0
 	param set MPC_Z_VEL_I 0.15
 	param set MPC_Z_VEL_P 0.6
+	param set MPC_XY_P 0.8
```
Comment (dagar) [inline ROMFS/px4fmu_common/init.d-posix/rcS]: These should be set in the appropriate airframe/model.

**Comment by dagar** [comment]
CI style check failure.

**Comment by bresch** [comment]
rebased on master

**Comment by bresch** [comment]
Replaced by #10746

**Comment by tuloski** [comment]
It would be nice to have a class C^4 position trajectory function for a quadrotor instead of C^2 as in this PR, i.e. to have up to a continuous jounce (or snap, the derivative of jerk). That is because of the dynamics of a multirotor. You have 4 "integrators" (2 in position and two in attitude) so the relative degree is 4. That is why you should have a C^4 function.
Also the trajectory derivatives (velocity, acceleration, jerk and snap) should be used as feedforward in the control loop to have a huge increase in tracking performance.

**Comment by bresch** [comment]
@tuloski I agree, the limitation is that it gets really tricky to analytically solve a time-optimal with arbitrary initial acceleration and arbitrary final velocity for equations of a degree higher than 2 to run and run the solver in real-time.
If you want to do it, please proceed. You can add a new flight task and have a try :)
Yes, the plan is to use the acceleration trajectory as a feedforward as soon as I solve a small chattering issue...

**Pull request closed without merging**